src/harness/reference_models/iap/iap.py

```diff
@@ -0,0 +1,654 @@
+    Updates global container aggregate_interference dictionary in the format of 
+    {(latitude,longitude,low_freq,high_freq,entity_type):aggregate_interference}
+    aggregate_interference value is in mW/RBW.
+    and global container post_iap_managed_grants nested dictionary in the format of
+    {grant_id:{ProtectionConstraint:PostIAPGrantInfo,ProtectionConstraint:PostIAPGrantInfo},
+    where ProtectionConstraint and PostIAPGrantInfo is a namedtuple 
+  """
+
+  for channel in channels:
+    # Set interference quota for protection_point, channel combination
+
+    # Calculate Roll of Attenuation for ESC Sensor Protection entity
+    if protection_ent_type is interf.ProtectedEntityType.ESC_CAT_A or\
+      protection_ent_type is interf.ProtectedEntityType.ESC_CAT_B:
+      if channel[0] >= 3650.e6:
+        center_freq = (channel[0] + channel[1]) / 2
+        interference_quota = interf.dbToLinear(interf.linearToDb(threshold) - 
+                (2.5 + ((center_freq - interf.ESC_CH21_CF_HZ) / interf.ONE_MHZ)))
+      else:
+        interference_quota = threshold
+
+    # Get the FSS blocking passband range
+    elif protection_ent_type is interf.ProtectedEntityType.FSS_BLOCKING:
+      interference_quota = threshold
+      channel = [low_freq, high_freq]
+
+    else:
+      interference_quota = threshold
+
+    # Get protection constraint over 5MHz channel range
+    channel_constraint = interf.ProtectionConstraint(latitude=protection_point[0],
+                           longitude=protection_point[1], low_frequency=channel[0],
+                           high_frequency=channel[1], entity_type=protection_ent_type)
+
+    # Identify CBSD grants in the neighborhood of the protection point 
+    # and channel
+    neighborhood_grants = interf.findOverlappingGrantsInsideNeighborhood(
+                            grant_objects, channel_constraint)
+    num_grants = len(neighborhood_grants)
+
+    if num_grants is 0:
+      continue
+
+    # calculate the fair share
+    fair_share = interference_quota / num_grants
+
+    # Empty list of final IAP EIRP and interference assigned to the grant
+    grants_eirp = []
+    interference_list = []
+    # Initialize interference list to 0 for all the neighborhood_grants
+    interference_list[:num_grants] = [0] * num_grants
+
+    aggr_interference = 0
+
+    # Initialize flag to False for all the grants
+    grant_initialize_flag = []
+    grant_initialize_flag[:num_grants] = [False] * num_grants
+
+    # Initialize grant EIRP to maxEirp for all the grants 
+    grants_eirp = [grant.max_eirp for grant in neighborhood_grants]
+
+    while (num_grants > 0):
+      for index, grant in enumerate(neighborhood_grants):
+        if grant_initialize_flag[index] is False:
+          # Compute interference grant causes to protection point over channel
+
+          # Compute interference to FSS Co-channel protection constraint
+          if protection_ent_type is interf.ProtectedEntityType.FSS_CO_CHANNEL:
```

Comment:
Why not having a entry function in interference.py which does this logic, a bit like you do with the findOverlappingGrantsInsideNeighborhood().
This will make the IAP logic more simple, as it just deal with the IAP algorithm.
Added benefit: allows for easy unit testing as you can mock that function, and test this function in isolation for proper working.
It will also make the vectorization of the loops easier, if we deem this necessary for speed.

Reply:
Introduced   new function computeInterference in interference.py